### PR TITLE
Fix bug for history requests with custom symbols

### LINKED ISSUE #8259: Problem retrieving CSV custom symbol history
#### Expected Behavior
Retrieve the quote history in a dataframe of a local CSV file on research notebook.

#### Actual Behavior
The dataframe is returning empty.

#### Potential Solution
The research docker image 16484 works fine, the bug started on the docker image 16487.
This [PR](https://github.com/QuantConnect/Lean/commit/621d1252b9c221a6442b11b084fdfb2e815da2cd) changing the `QCAlgorihtm.Python.History()` is from the same day of the 16487 image so maybe it introduced the bug.

#### Reproducing the Problem
Create the file `test1.csv` in `Lean/data` folder, fill it with some data rows.
Using the 16487 image and so on the `history` dataframe returns empty.

```
from datetime import datetime

from AlgorithmImports import *


class TestTradeBar(TradeBar):
    def get_source(self, config: SubscriptionDataConfig, date: datetime, is_live_mode: bool) -> SubscriptionDataSource:
        return SubscriptionDataSource(source=f'{Globals.data_folder}/test1.csv',
                                      transportMedium=SubscriptionTransportMedium.LOCAL_FILE,
                                      format=FileFormat.CSV)

    def reader(self, config: SubscriptionDataConfig, line: str, date: datetime, is_live_mode: bool) -> BaseData:
        if not line[0].isdigit():
            return None
        data = line.split(',')
        bar_time = datetime.utcfromtimestamp(int(data[0]))

        open = float(data[1])
        high = float(data[2])
        low = float(data[3])
        close = float(data[4])
        volume = int(float(data[7]))
        return TradeBar(bar_time, config.symbol, open, high, low, close, volume)


qb = QuantBook()
qb.add_data(
    type=TestTradeBar,
    ticker='TEST1',
    properties=SymbolProperties(
        description='TEST1',
        quoteCurrency='USD',
        contractMultiplier=1,
        minimumPriceVariation=0.01,
        lotSize=1,
        marketTicker='TEST1',
    ),
    exchange_hours=SecurityExchangeHours.always_open(TimeZones.NEW_YORK),
    resolution=Resolution.MINUTE,
    fill_forward=True,
    leverage=1,
)
history = qb.history(qb.securities.keys(), datetime(2024, 8, 2), datetime(2024, 8, 3))
history
```

### PR DESCRIPTION
**Pull request by Marinovsky**
#### Description

When the symbol used for the history request was custom, even though the data was obtained, it was discarded in the method `PandasConverter.AddSliceDataTypeDataToDict()`. This happened since `PandasConverter.GetPandasDataValue` considered a given data to be custom only if the data type was custom. Therefore, if the custom symbol inherited from Tradebar, it wasn't considered as custom data. This PR aims to solve that bug and centralize an existent method that determines if a symbol is custom or not.
#### Related Issue

Closes #8259 
#### Motivation and Context

With this change users will be able to request history data for custom symbols in research notebooks
#### Requires Documentation Change

N/A
#### How Has This Been Tested?

I used the algorithm reproducing the bug from the PR to make a unit test that asserted the bug was solved.
#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Algorithm/QCAlgorithm.History.cs [hunk 1]
```diff
@@ -949,7 +949,7 @@ protected IEnumerable<HistoryRequest> CreateDateRangeHistoryRequests(IEnumerable
             DataNormalizationMode? dataNormalizationMode = null, int? contractDepthOffset = null)
         {
             symbols = symbols.ToArray();
-            return CreateDateRangeHistoryRequests(symbols, GetCustomDataTypeFromSymbols((Symbol[])symbols) ?? typeof(BaseData), startAlgoTz, endAlgoTz, resolution, fillForward, extendedMarketHours,
+            return CreateDateRangeHistoryRequests(symbols, Extensions.GetCustomDataTypeFromSymbols((Symbol[])symbols) ?? typeof(BaseData), startAlgoTz, endAlgoTz, resolution, fillForward, extendedMarketHours,
```
Comment (jhonabreul) [inline Algorithm/QCAlgorithm.History.cs:952]: Is this cast to `Symbol[]` safe? The input `IEnumerable` might not be an array

File: Common/Python/PandasData.cs [hunk 2]
```diff
@@ -134,8 +135,8 @@ public PandasData(object data)
 
             var type = data.GetType();
             _isFundamentalType = type == typeof(Fundamental);
-            IsCustomData = type.Namespace != typeof(Bar).Namespace;
             _symbol = ((IBaseData)data).Symbol;
+            IsCustomData = type.Namespace != typeof(Bar).Namespace || Extensions.GetCustomDataTypeFromSymbols(new Symbol[] { _symbol }) != null ? true : false;
```
Comment (jhonabreul) [inline Common/Python/PandasData.cs:139]: 1. Only types in `QuantConnect.Data.Market` are cosidered "not custom"? What about `QuantConnect.Data`?
2. nit - the ternary operation might not be necessary: `!= null` is enough.
3. This check is also in `Algorithm/QCAlgorithm.History.cs:1087`: maybe move it to a single location like Extensions for single source of truth

**Review by jhonabreul — COMMENTED** [review]
Nice, ty!
Leaving some minor comments below.

File: Common/Python/PandasData.cs (hunk 2 above)
Comment (Marinovsky) [reply Common/Python/PandasData.cs:139]: 1. The other data types present in QuantConnect.Data are `BaseData` and `DynamicData`, but they are abstract, so we cannot instantiate an object of those classes. Then, it seems that the only custom types are the ones not present in `QuantConnect.Data.Market`.

**Review by jhonabreul — APPROVED** [review]
Thank you!

**Review by Martin-Molinero — APPROVED** [review]
🚀

**Pull request merged**